src/tests/stuf/test_message.py

```diff
@@ -0,0 +1,127 @@
+
+        message = StufMessage(msg)
+        self.assertIsNone(message.namespaces)
+        mock_set_namespaces.assert_called_with(msg)
+
+
+@patch("gobstuf.stuf.message.StufMessage.load", MagicMock())
+class StufMessageTest(TestCase):
+
+    @patch("gobstuf.stuf.message.StringIO")
+    @patch("gobstuf.stuf.message.ET")
+    def test_set_namespaces(self, mock_et, mock_stringio):
+        mock_et.iterparse.return_value = [
+            (1, ('prefix1', 'url1')),
+            (2, ('prefix2', 'url2')),
+            (3, ('prefix3', 'url3')),
+        ]
+        msg = 'msg'
+        mock_stringio.side_effect = lambda x: 'stringio(' + x + ')'
+
+        message = StufMessage('')
+        message.set_namespaces(msg)
+        mock_et.iterparse.assert_called_with('stringio(msg)', events=['start-ns'])
+        mock_et.register_namespace.assert_has_calls([
+            call('prefix1', 'url1'),
+            call('prefix2', 'url2'),
+            call('prefix3', 'url3'),
+        ])
+        self.assertEqual({
+            'prefix1': 'url1',
+            'prefix2': 'url2',
+            'prefix3': 'url3',
+        }, message.namespaces)
+
+    def test_find_elm(self):
+        message = StufMessage('')
+        message.tree = MagicMock()
+        message.namespaces = MagicMock()
+
+        res = message.find_elm('a b c')
+
+        # Recursive return value
+        self.assertEqual(message.tree.find().find().find(), res)
+
+        # Recursive calls
+        message.tree.find.assert_has_calls([
+            call('a', message.namespaces),
+            call().find('b', message.namespaces),
+            call().find().find('c', message.namespaces),
+        ])
+
+    def test_set_elm_value(self):
+        message = StufMessage('')
+        message.find_elm = MagicMock()
+        message.tree = MagicMock()
+        message.set_elm_value('some element', 'some value')
+        message.find_elm.assert_called_with('some element', None)
+        self.assertEqual('some value', message.find_elm().text)
+
+        mocked_tree = MagicMock()
+        message.set_elm_value('some element', 'some value', mocked_tree)
+        message.find_elm.assert_called_with('some element', mocked_tree)
+
+    def test_get_elm_value(self):
+        message = StufMessage('')
+        message.find_elm = MagicMock()
+        message.tree = MagicMock()
+        res = message.get_elm_value('some element')
+        message.find_elm.assert_called_with('some element', None)
+        self.assertEqual(message.find_elm().text, res)
+
+        mocked_tree = MagicMock()
+        message.get_elm_value('some element', mocked_tree)
+        message.find_elm.assert_called_with('some element', mocked_tree)
+
+    @patch("gobstuf.stuf.message.ET")
+    def test_to_string(self, mock_et):
+        message = StufMessage('')
+        message.tree = MagicMock()
+
+        self.assertEqual(mock_et.tostring(), message.to_string())
+        mock_et.tostring.assert_called_with(message.tree, encoding='unicode')
+
```

Comment:
Basically you test the implementation not the behaviour. The result is expected to be a unicoded string. How this is done is the implementation. If you want to test the behaviour then you should construct a tree with some text that is supposed to be returned as a unicoded string. How it is done is up to the method.
It is basically done in the whole test of message.py, the tests assert that the implementation is as it is, not that the behaviour is as it should be.
Well this is a personal comment...

Reply:
Yes correct ;)

However, I think this is correct. The tested method performs 1 call to an external library that is mocked. I'm assuming the external library does what it says on the tin. With that assumption I'm testing if the code is calling that external library correctly.

I know there's a ET.tostring() method, I know my parameter self.tree is set correctly (in a different tested method). If I know my parameters are correct and I'm using the external method correctly, then I'm good.

Of course, if the external code changes my code might stop working, but I prefer to see that as a calculated risk; I like to mock everything that's not in my control, to keep the tests clear and focused. Personal preference I guess ;)